test/utils.cpp

```diff
@@ -589,6 +592,47 @@ TEST_CASE("run_program()", "[utils]")
 	REQUIRE(utils::run_program(argv, "") == "hello world");
 }
 
+TEST_CASE("run_program() works for large inputs", "[utils]")
+{
+	const std::string large_input(1000000, 'a');
+
+	// Calling std::thread::~thread() on a thread that's still running will
+	// make it call `std::terminate`, crashing our entire test suite. To avoid
+	// that, this wrapper kills the thread with SIGKILL and detaches it from
+	// the thread handle.
+	struct Runner {
+		std::thread runner;
+
+		Runner(std::thread runner)
+			: runner(std::move(runner))
+		{}
+
+		~Runner()
+		{
+			pthread_kill(runner.native_handle(), SIGKILL);
+			runner.detach();
+		}
+	};
+
+	std::atomic_flag thread_finished = ATOMIC_FLAG_INIT;
+	std::string result;
+	Runner runner(std::thread([large_input, &thread_finished, &result]() {
+		const char* argv[4];
+		argv[0] = "sh";
+		argv[1] = "-c";
+		argv[2] = "cat";
+		argv[3] = nullptr;
+		result = utils::run_program(argv, large_input);
+		thread_finished.test_and_set();
+	}));
+
+	// cat should be able to process 1MB of input in under a second
+	std::this_thread::sleep_for(std::chrono::seconds(1));
```

Comment:
Alternatively, we could use something like a condition_variable and use its [wait_for](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_for) method to only wait the minimal amount of time for the thread to finish while specifying a timeout to catch the old error behavior.
However, 1 second is not a problem, and the current implementation is probably easier to read/understand, so I'm fine keeping it as-is.

Reply:
That's a good idea since our C++ tests are single-threaded. Done: https://github.com/newsboat/newsboat/pull/2511/commits/b9b0855f9832168d9ea2dd653adeba1717918a67